File: Java/src/net/kenyang/algorithm/BestTimeToBuyAndSellStock.java
package net.kenyang.algorithm;

/**
 * Say you have an array for which the i element is the price of a given stock on day i.
 * If you were only permitted to complete at most one transaction (ie, buy one and sell one share of the stock), design an algorithm to find the maximum profit.
 * @author Ken Yang
 *
 */
public class BestTimeToBuyAndSellStock {
    public int maxProfit(int[] prices) {
        int size = prices.length;
        
        if (size<=1){
            return 0;
        }

        int min = prices[0];
        int max = prices[1] - min;
        for (int i = 2; i < size; i++) {
            
            if (min > prices[i-1]) {
                min = prices[i-1];
            }
            
            if (max < (prices[i]-min)) {
                max = prices[i]-min;
            }
            
        }
        
        
        return (max<0) ? 0 : max;
    }
}

File: Python/bestTimeToBuyAndSellStock.py
# Time: O(n)
# Space: O(1)
#
# Say you have an array for which the ith element 
# is the price of a given stock on day i.
#
# If you were only permitted to complete at most one transaction 
# (ie, buy one and sell one share of the stock), 
# design an algorithm to find the maximum profit.

class Solution:
    # @param prices, a list of integer
    # @return an integer
    def maxProfit(self, prices):
        max_profit, min_price = 0, 9223372036854775807
        for price in prices:
            min_price = min(min_price, price)
            max_profit = max(max_profit, price - min_price)  
        return max_profit

if __name__ == "__main__":
    result = Solution().maxProfit([3, 2, 1, 4, 2, 5, 6])
    print result
    